tests/php/Control/HTTPRequestTest.php

```diff
@@ -7,11 +7,36 @@
 use SilverStripe\Control\Middleware\TrustedProxyMiddleware;
 use SilverStripe\Control\Session;
 use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Control\Tests\HTTPRequestTest\HTTPRequestTestException;
+use PHPUnit\Framework\Attributes\DataProvider;
 
 class HTTPRequestTest extends SapphireTest
 {
     protected static $fixture_file = null;
 
+    /**
+     * @param callable|null $oldHandler
+     */
+    private $oldHandler = null;
+
+    protected function setup(): void
+    {
+        // Use custom error handler to allow the use of expectException() to catch warnings
+        // which is required in testWildCardWithFurtherParams().
+        // PHPUnit does not support expecting warnings
+        parent::setup();
+        $this->oldHandler = set_error_handler(function (int $errno, string $errstr, string $errfile, int $errline) {
+            throw new HTTPRequestTestException($errstr);
+        });
+    }
+
+    protected function tearDown(): void
+    {
+        restore_error_handler();
+        $this->oldHandler = null;
+        parent::tearDown();
+    }
```

Comment:
This is a bad pattern, because it means any time we want to `expectException` we have to do it in this decentralised way. It would be easy to do this in 20 different ways across 20 different tests and means reimplementing or at least copying the logic every time we do it.

Implement a `expectWarning` method in `SapphireTest` instead, to mimic the old API that `TestCase` used to have.

Reply:
Same with any other similar changes (probably there's `expectError` and `expectDeprecation` somewhere as well)

Reply:
Updated